### ISSUE TITLE
Case study: local-only WiFi password QR code generator

### ISSUE THREAD

**Issue by tiziano88** [project-oak/oak#1529]
Spawned from https://github.com/project-oak/oak/issues/1492 .

Case study:

Random developer wants to implement an app to generate a QR code from a WiFi password.

Looking around, there are currently a number of such apps available, including web-based and native, e.g.:

- https://qifi.org/
- https://www.qr-code-generator.com/solutions/wifi-qr-code/
- https://qrcode.tec-it.com/en/wifi
- https://coda.io/@rod/wifi-qr-code-generator

Some of them have some text trying to justify that it is safe to enter passwords there, for instance, from https://qifi.org/
:

> About
Ever wanted to create a cool QR code for your guests? But never wanted to type in your WiFi credentials into a form that submits them to a remote webserver to render the QR code? QiFi for the rescue! It will render the code in your browser, on your machine, so the WiFi stays as secure as it was before (read the code if you do not trust text on the internet :-))!
>
> If you use the Save-button to store a code, this is still secure, as the data is stored in HTML5 localStorage and is never transmitted to the server (in contrast to cookie-based solutions).
>
> Don't trust your browser either? Just pipe the string WIFI:S:<SSID>;T:<WPA|WEP|>;P:<password>;; through the QR code generator of your choice after reading the documentation.

User wants to use such app but because because it requires entering their own WiFi password, they want to be sure that such app cannot send anything over to a remote server.

Note that this is not even specific to web apps; we do this all the time by downloading binaries (e.g. games, text editors, etc.) and running them locally, under the assumption that they behave correctly. In some cases it is possible to download the source and either inspect it or compile it locally in order to have some guarantees, but in practice this tends not to happen, and if it does happen requires inspecting the entirety of the code (including third party dependencies).

I would like to think more how this example, which is arguably much simpler than many other Oak use cases, can actually be solved.

In particular, it seems that this should be solved completely locally, and with no reliance on a remote server or remote attestation between client and server.

If we imagine that the logic (including rendering) of the app is contained in a single Wasm file, then this seems solvable by introducing a web-based version of the Oak Runtime which is hosted on a well-known domain (e.g. https://oak.google.com) which is trusted **by assumption** by users once and for all (i.e. does not depend on the particular implementation of the third party app that will run on top of it). 

For instance, the Oak host page may contain a bit of logic roughly equivalent to the Oak Runtime, plus additional logic for rendering DOM efficiently.

A third party app like the WiFi password QR code generator may then be implemented as a Wasm module that relies on the Oak Wasm ABI; in particular it would emit messages on an Oak channel describing a DOM for the app, e.g. describing a structure with a text box and a button, attaching identifiers used to route messages back into the app, and accept such messages back on another Oak channel (which is a common pattern for modern event-loop based web application frameworks, e.g. https://github.com/yewstack/yew , https://guide.elm-lang.org/architecture/ , https://reactjs.org/).

When loaded on the trusted oak loader page, the loader would then reserve a portion of its own DOM for the app to render its own content (e.g. a div in the middle of the page), while also policing the allowed and disallowed DOM nodes (e.g. removing any references or links to external resources, or disallowing scripts).

To start with, the trusted web Oak loader would just disallow any communication to remote servers and enforce a "local-only policy" for a single-node Wasm application.

In the future, the web Oak loader would support communication with back-ends, as long as they are Oak-enabled, and allow the user to inspect and modify labels attached to these outgoing requests.

cc @project-oak/core

**Comment by juliettepretot** [comment]
Thanks for creating this usecase. :-) To clarify, it seems this depends on #1067? In the example you outlined it seems that the entire Oak application would run in the browser itself, and only later on would connections to remote oak servers become possible.

**Comment by juliettepretot** [comment]
I know this is a simple use-case as a start, but in my mind Oak seems a little bit overkill for on-device processing use-cases. 

This goes a bit back to @seefeldb [comment](https://github.com/project-oak/oak/issues/1492#issuecomment-700185832) in #1492, Oak & IFCs only really becomes uniquely useful when data aggregation etc occurs. Note that this doesn't discount Oak functionality to render web UI, but in my mind it'll only really shine when connecting to remote oak servers.

**Comment by tiziano88** [comment]
Yes this assumes the existence of a web-based Oak emulator of some sort, and this specific use case seems solvable by a client-only application, i.e. no network communication required.

Of course Oak would be overkill for something like this, but my point is that if Oak cannot satisfyingly support (and explain correctly to users) something this simple, in which there are no enclaves, no network communication, no delegation, etc., then there is not much point in thinking about very complex examples (or maybe there is a point in thinking about them in parallel; anyways, we should not skip the simple cases).

**Comment by juliettepretot** [comment]
> Of course Oak would be overkill for something like this, but my point is that if Oak cannot satisfyingly support (and explain correctly to users) something this simple, in which there are no enclaves, no network communication, no delegation, etc., then there is not much point in thinking about very complex examples (or maybe there is a point in thinking about them in parallel; anyways, we should not skip the simple cases).

Fair point. :)) I guess it doesn't make for a good roadshow type of example to demonstrate Oak's power, but agreed we should be able to handle it.

**Comment by tiziano88** [comment]
I do think this is a genuinely unsolved problem though, regardless of whether Oak is overkill for it or not. I think @seefeldb suggested that this seems solvable by iframes with sandboxing and CSP headers, but I didn't manage to hack something together that solves it, at least.

https://jsfiddle.net/02y3r64m/11/ is my best attempt so far, and it does not actually convincingly solve the problem of displaying / embedding a web page with a guarantee that it cannot just send whatever I type into it to any remote server. I'd be curious to learn more about this though, and whether it is actually solvable nicely with existing web technologies.

**Comment by juliettepretot** [comment]
You cannot enforce this via CSP, for the same reasons we abandoned the chrome extension. You won't be able to block redirects etc. 

There is a proposal to add JavaScript "realms", which would potentially allow for the execution of arbitrary JS code while making sure it cannot leak data. https://github.com/tc39/proposal-realms/blob/main/explainer.md

There's also web workers for running JS code in isolation, but they include their own API for making network calls. https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers

For any of these (flawed) approaches though you wouldn't be able to embed a whole website, but rather run JS code in isolation. You'd still need some API to allow for limited access to the DOM to display UI, similar to our Oak approach. 

Indeed there is no standards based way to make this happen on the web right now.

**Comment by juliettepretot** [comment]
If I had to solve this problem right now, I would likely resort to writing a script deno that accepts the the password as a flag and outputs the QR code to the CLI. Admittedly it wouldn't include any UI. 

It's permissions system would prohibit network access. https://deno.land/manual/getting_started/permissions

**Comment by juliettepretot** [comment]
The simplest way to solve this in the browser without having to resort to Oak & WASM imho would to use web workers, and overwrite the global object to remove methods to leak data. 

Wrote a small example, which will execute arbitrary JS and log the result to the console.
https://codesandbox.io/s/static-forked-y7n1z

For example passing it an encoded function that adds 1 + 1: https://y7n1z.csb.app/?functionBody=return%201%20%2B%201

Just like our Oak solution though, to render UI returned by abitrary code this would require a library to render UI without exposing the ability to inject scripts. That's possible, but requires a bit more work.

**Comment by juliettepretot** [comment]
Going back to the discussion with @seefeldb in #1492 this would be possible to achieve via CSP, a web host that would serve a website with these CSPs, thereby forbidding navigation & leakage: `Content-Security-Policy: default-src 'none'; sandbox allow-scripts;`

**Comment by ipetr0v** [comment]
> To start with, the trusted web Oak loader would just disallow any communication to remote servers and enforce a "local-only policy" for a single-node Wasm application.

> I think @seefeldb suggested that this seems solvable by iframes with sandboxing and CSP headers

I have a couple of questions about, how it will prohibit external communications and only allow data to be sent to Oak.
IIUC:
- VDOM returned by the Oak application will be placed in an `iframe` that resides in the original DOM
- VDOM in the `iframe` is not allowed to have JavaScript

So the only way to send data from the client is to have something like:
```html
<a class="button" href="/something/page">Send Wi-Fi</a>
```
Questions:
- Is it possible to prohibit sending data everywhere except for one place - inside Oak
- How should `href` look like in order to send data to Oak, that resides in the parental DOM (related to the `iframe`)

**Comment by tiziano88** [comment]
@ipetr0v this use case is specifically about no network communication: everything happens locally in the browser.

The question is how to convincingly prove that this is actually the case to a user.

**Comment by juliettepretot** [comment]
A chrome extension could verify that the correct CSPs are set to forbid navigation & network connectivity.

**Comment by ipetr0v** [comment]
IIUC, even if the network is prohibited, the new iframed DOM can only send HTTP requests (since we also prohibit JavaSctipt (do we?)). So where will this DOM send HTTP requests to? Is it possible to send them locally to the same page outside the `iframe`?

**Comment by tiziano88** [comment]
The point is that it must be prevented from sending HTTP requests in the first place.

**Comment by ipetr0v** [comment]
My question is about, how it will communicate with the web-based runtime, if JS is disabled?

**Comment by juliettepretot** [comment]
JS won't be disabled. :)

**Comment by ipetr0v** [comment]
Then I might understood it wrong. I was thinking that this new `iframe` DOM would be just a plain HTML without JS =)

**Comment by juliettepretot** [comment]
Naw, the idea of the iframe was to use it to forbid leaking data via navigating to another page. :) (The same can be achieved via CSPs too)
